File: Frontend/App/taskforge/lib/services/project_service.dart
// lib/services/project_service.dart
import 'dart:io';
import 'package:graphql_flutter/graphql_flutter.dart';
import '../api/graphql_client.dart';
import '../api/rest_client.dart';
import '../models/project.dart';

class ProjectService {
  static final GraphQLClient _graphQLClient = GQL.client.value;

  // GraphQL operations for main CRUD
  static Future<List<Project>> getProjects() async {
    final QueryOptions options = QueryOptions(
      document: gql(GraphQLQueries.getProjects),
      fetchPolicy: FetchPolicy.cacheAndNetwork,
    );

    final QueryResult result = await _graphQLClient.query(options);

    if (result.hasException) {
      throw Exception('Failed to fetch projects: ${result.exception.toString()}');
    }

    final List<dynamic> projectsData = result.data?['projects'] ?? [];
    return projectsData.map((projectJson) => Project.fromJson(projectJson)).toList();
  }

  static Future<Project> createProject(CreateProjectInput input) async {
    final MutationOptions options = MutationOptions(
      document: gql(GraphQLMutations.createProject),
      variables: {'input': input.toJson()},
    );

    final QueryResult result = await _graphQLClient.mutate(options);

    if (result.hasException) {
      throw Exception('Failed to create project: ${result.exception.toString()}');
    }

    return Project.fromJson(result.data?['createProject']);
  }

  static Future<Project> updateProject(String id, Map<String, dynamic> updates) async {
    final MutationOptions options = MutationOptions(
      document: gql(GraphQLMutations.updateProject),
      variables: {
        'id': id,
        'input': updates,
      },
    );

    final QueryResult result = await _graphQLClient.mutate(options);

    if (result.hasException) {
      throw Exception('Failed to update project: ${result.exception.toString()}');
    }

    return Project.fromJson(result.data?['updateProject']);
  }

  static Future<bool> deleteProject(String id) async {
    final MutationOptions options = MutationOptions(
      document: gql(GraphQLMutations.deleteProject),
      variables: {'id': id},
    );

    final QueryResult result = await _graphQLClient.mutate(options);

    if (result.hasException) {
      throw Exception('Failed to delete project: ${result.exception.toString()}');
    }

    return result.data?['deleteProject'] == true;
  }

  // Analytics operations using GraphQL (Prisma data)
  static Future<ProjectAnalytics?> getProjectAnalytics(String projectId) async {
    final QueryOptions options = QueryOptions(
      document: gql(GraphQLQueries.getProjectAnalytics),
      variables: {'projectId': projectId},
      fetchPolicy: FetchPolicy.cacheAndNetwork,
    );

    final QueryResult result = await _graphQLClient.query(options);

    if (result.hasException) {
      throw Exception('Failed to fetch project analytics: ${result.exception.toString()}');
    }

    final analyticsData = result.data?['projectAnalytics'];
    return analyticsData != null ? ProjectAnalytics.fromJson(analyticsData) : null;
  }

  // File operations using REST API
  static Future<Map<String, dynamic>> exportProject(String projectId) async {
    try {
      return await RestClient.exportProject(projectId);
    } catch (e) {
      throw Exception('Failed to export project: $e');
    }
  }

  static Future<Map<String, dynamic>> importProject(File file) async {
    try {
      return await RestClient.importProject(file);
    } catch (e) {
      throw Exception('Failed to import project: $e');
    }
  }

  // Real-time updates using Stream
  static Stream<List<Project>> watchProjects() {
    final WatchQueryOptions options = WatchQueryOptions(
      document: gql(GraphQLQueries.getProjects),
      fetchPolicy: FetchPolicy.cacheAndNetwork,
      pollInterval: const Duration(seconds: 60), // Poll every minute
    );

    final ObservableQuery observableQuery = _graphQLClient.watchQuery(options);
    
    return observableQuery.stream.map((result) {
      if (result.hasException) {
        throw Exception('Failed to watch projects: ${result.exception.toString()}');
      }

      final List<dynamic> projectsData = result.data?['projects'] ?? [];
      return projectsData.map((projectJson) => Project.fromJson(projectJson)).toList();
    });
  }

  // Offline support
  static Future<List<Project>> getCachedProjects() async {
    final QueryOptions options = QueryOptions(
      document: gql(GraphQLQueries.getProjects),
      fetchPolicy: FetchPolicy.cacheOnly,
    );

    final QueryResult result = await _graphQLClient.query(options);

    if (result.hasException || result.data == null) {
      return [];
    }

    final List<dynamic> projectsData = result.data?['projects'] ?? [];
    return projectsData.map((projectJson) => Project.fromJson(projectJson)).toList();
  }

  // Project member management
  static Future<Project> addMember(String projectId, String userId) async {
    return await updateProject(projectId, {
      'addMemberIds': [userId],
    });
  }

  static Future<Project> removeMember(String projectId, String userId) async {
    return await updateProject(projectId, {
      'removeMemberIds': [userId],
    });
  }

  static Future<Project> updateProjectRole(String projectId, String userId, String role) async {
    return await updateProject(projectId, {
      'memberRoles': {userId: role},
    });
  }
}

File: Frontend/Web/src/services/project.service.js
import api from '@/lib/api'

export class ProjectService {
  /**
   * Get all projects for the current user
   * @returns {Promise<Array>}
   */
  static async getAllProjects() {
    const response = await api.get('/projects/all')
    return response.data
  }

  /**
   * Get a single project by ID
   * @param {string} projectId
   * @returns {Promise<Object>}
   */
  static async getProject(projectId) {
    const response = await api.get(`/projects/${projectId}`)
    return response.data
  }

  /**
   * Create a new project
   * @param {Object} projectData
   * @returns {Promise<Object>}
   */
  static async createProject(projectData) {
    const response = await api.post('/projects/create', projectData)
    return response.data
  }

  /**
   * Update an existing project
   * @param {string} projectId
   * @param {Object} updates
   * @returns {Promise<Object>}
   */
  static async updateProject(projectId, updates) {
    const response = await api.put(`/projects/${projectId}`, updates)
    return response.data
  }

  /**
   * Delete a project
   * @param {string} projectId
   * @returns {Promise<Object>}
   */
  static async deleteProject(projectId) {
    const response = await api.delete(`/projects/${projectId}`)
    return response.data
  }

  /**
   * Add a member to a project
   * @param {string} projectId
   * @param {Object} memberData - { userId, role }
   * @returns {Promise<Object>}
   */
  static async addMember(projectId, memberData) {
    const response = await api.post(`/projects/${projectId}/members`, memberData)
    return response.data
  }

  /**
   * Remove a member from a project
   * @param {string} projectId
   * @param {string} userId
   * @returns {Promise<Object>}
   */
  static async removeMember(projectId, userId) {
    const response = await api.delete(`/projects/${projectId}/members/${userId}`)
    return response.data
  }

  /**
   * Get project analytics
   * @param {string} projectId
   * @returns {Promise<Object>}
   */
  static async getProjectAnalytics(projectId) {
    const response = await api.get(`/projects/${projectId}/analytics`)
    return response.data
  }

  /**
   * Archive a project
   * @param {string} projectId
   * @returns {Promise<Object>}
   */
  static async archiveProject(projectId) {
    const response = await api.patch(`/projects/${projectId}/archive`)
    return response.data
  }

  /**
   * Restore an archived project
   * @param {string} projectId
   * @returns {Promise<Object>}
   */
  static async restoreProject(projectId) {
    const response = await api.patch(`/projects/${projectId}/restore`)
    return response.data
  }
}

export default ProjectService
